smithy-swift-codegen/src/main/kotlin/software/amazon/smithy/swift/codegen/integration/HttpProtocolUnitTestResponseGenerator.kt

```diff
@@ -139,44 +138,97 @@ open class HttpProtocolUnitTestResponseGenerator protected constructor(builder:
         }
     }
-}
 
-fun renderMemberAssertions(writer: SwiftWriter, test: HttpMessageTestCase, members: Collection<MemberShape>, model: Model, symbolProvider: SymbolProvider, expected: String, actual: String) {
-    for (member in members) {
-        val shape = model.expectShape(member.target.toShapeId())
-        val baseVarName = symbolProvider.toMemberName(member)
-        val expectedMemberName = "$expected.$baseVarName"
-        val actualMemberName = "$actual.$baseVarName"
-        val suffix = if (symbolProvider.toSymbol(member).isBoxed()) "?" else ""
-        if (member.isStructureShape) {
-            writer.write("XCTAssert(\$L === \$L)", expectedMemberName, actualMemberName)
-        } else if ((shape.isDoubleShape || shape.isFloatShape)) {
-            val stringNodes = test.params.stringMap.values.map { it.asStringNode().orElse(null) }
-            if (stringNodes.isNotEmpty() && stringNodes.mapNotNull { it?.value }.contains("NaN")) {
-                writer.write("XCTAssertEqual(\$L$suffix.isNaN, \$L$suffix.isNaN)", expectedMemberName, actualMemberName)
-            } else {
-                writer.write("XCTAssertEqual(\$L, \$L)", expectedMemberName, actualMemberName)
+    private fun identifier(ctx: ProtocolGenerator.GenerationContext, shape: Shape): String {
+        return "${ctx.service.id}.${shape.id}"
+    }
+
+    private fun renderEquatable(shape: Shape) {
+        if (!(shape is StructureShape || shape is UnionShape)) { return }
+        val symbol = ctx.symbolProvider.toSymbol(shape)
+        val httpBindingSymbol = Symbol.builder()
+            .definitionFile("./${ctx.settings.moduleName}Tests/models/${symbol.name}+Equatable.swift")
+            .name(symbol.name)
+            .build()
+        ctx.delegator.useShapeWriter(httpBindingSymbol) { writer ->
+            writer.addImport(ctx.settings.moduleName)
+            writer.openBlock("extension \$L: \$N {", "}", symbol.fullName, SwiftTypes.Protocols.Equatable) {
+                writer.write("")
+                writer.openBlock("public static func ==(lhs: \$L, rhs: \$L) -> Bool {", "}", symbol.fullName, symbol.fullName) {
+                    when (shape) {
+                        is StructureShape -> {
+                            shape.members().filter { !it.hasTrait<HttpQueryTrait>() }.forEach { member ->
+                                val propertyName = ctx.symbolProvider.toMemberName(member)
+                                val path = "properties.".takeIf { shape.hasTrait<ErrorTrait>() } ?: ""
+                                val propertyAccessor = "$path$propertyName"
+                                val target = ctx.model.expectShape(member.target)
+                                when (target) {
+                                    is FloatShape, is DoubleShape -> {
+                                        writer.write(
+                                            "if (lhs.\$L?.isNaN ?? false && rhs.\$L?.isNaN ?? false) { return true }",
```

Comment:
Just realized that if you `return true` after this check, it will short-circuit whatever checks remain on the structure's fields.  So either put it back the way it was, or find another way to get the required behavior.

Reply:
Fixed via helper function.